### DISCUSSION TITLE
Add customized particle Informations

### DISCUSSION THREAD

**Discussion by Perpetual3134** [InteractiveComputerGraphics/SPlisHSPlasH#325] [Q&A]
Hello there,

i would like to add an additional parameter for each particle that is also exportet to the bgeos.
( one Int value contains a 0 or 1 or 2)

How can I implement this?

**Comment by janbender** [comment]
You can add a field for this parameter. See for example here: 

https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/dc2f63f3188cfe5a6fac8f6837de1335782958be/SPlisHSPlasH/FluidModel.cpp#L65C2-L65C118

Example: 
addField({ "id", FieldType::UInt, [&](const unsigned int i) -> unsigned int* { return &getParticleId(i); }, true });
removeFieldByName("id");

If you added a field, you can tell the exporter to export this field by setting the following parameter in the scene file: 

"particleAttributes": “velocity;id”

**Comment by Perpetual3134** [comment]
Thank you very much.

I was able to add a new field. I use a Vector3 type to store a color code.

`	addField({ "type", FieldType::Vector3, [&](const unsigned int i) -> Real* { return &getType(i)[0]; }, true });`

However, at the moment the program just crashes. Probably because I am not yet using the new field in the calculation.


The field should be used to distinguish the particles of the surface, the “second row” (blind particles) and the inner particles.

Now I am looking for a suitable place for my code to calculate the field. 

Would the place in the TimeStep.cpp at the end of the for-loop of line 96 be suitable?

**Comment by Perpetual3134** [comment]
Please, can you help?

**Comment by janbender** [comment]
The loop in line 96 is intended to compute the densities. I would not mix this with the code computation. Why not just write a new function and call it, e.g. in TimeStepDFSPH::step()?
